lib/spack/spack/schema/environment.py

```diff
@@ -0,0 +1,73 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""Schema for environment modifications. Meant for inclusion in other
+schemas.
+"""
+
+array_of_strings_or_num = {
+    'type': 'array', 'default': [], 'items':
+        {'anyOf': [{'type': 'string'}, {'type': 'number'}]}
+}
+
+dictionary_of_strings_or_num = {
+    'type': 'object', 'patternProperties':
+        {r'\w[\w-]*': {'anyOf': [{'type': 'string'}, {'type': 'number'}]}}
+}
+
+definition = {
+    'anyOf': [
+        {
+            'type': 'object',
+            'default': {},
+            'additionalProperties': False,
+            'properties': {
+                'set': dictionary_of_strings_or_num,
+                'unset': array_of_strings_or_num,
+                'prepend_path': dictionary_of_strings_or_num,
+                'append_path': dictionary_of_strings_or_num,
+                'remove_path': dictionary_of_strings_or_num
+            }
+        }, {
+            'type': 'object',
+            'default': {},
+            'additionalProperties': False,
+            'properties': {
+                'set': dictionary_of_strings_or_num,
+                'unset': array_of_strings_or_num,
+                'prepend-path': dictionary_of_strings_or_num,
+                'append-path': dictionary_of_strings_or_num
+            }
+        }
+    ]
+}
+
+
+def parse(config_obj):
+    """Returns an EnvironmentModifications object containing the modifications
+    parsed from input.
+
+    Args:
+        config_obj: a configuration dictionary conforming to the
+            schema definition for environment modifications
+    """
+    import spack.util.environment as ev
+    try:
+        from collections import Sequence  # novm
+    except ImportError:
+        from collections.abc import Sequence  # novm
+
+    env = ev.EnvironmentModifications()
+    for command, variable in config_obj.items():
+        # Ensures backward compatibility on both append-path
+        # and append_path (similarly for prepend-path vs. prepend_path)
+        command = command.replace('-', '_')
+        if isinstance(variable, Sequence):
```

Comment:
I don't fully understand this check. I think I've convinced myself it's correct from your tests, but can you add a comment to explain why we're checking for a sequence here?

Reply:
The check is needed to distinguish `unset`, which takes only a list of variables names, from the other commands that require a mapping from variable names to the value used to modify them. I'll add a comment to make that clear.

Reply:
Comment added